### DISCUSSION TITLE
Can We Ping It?

### DISCUSSION THREAD

**Discussion by sethp** [rustbox/discussion#3] [General]
Tracking notes/discussion thread for the goal: "Can we run Chipyard's [pingd test](https://github.com/ucb-bar/chipyard/blob/main/tests/pingd.c)?"

- [x] Can we ping a simulated SoC?
- [ ] Can we ping an FPGA? Maybe one of [Diligent's Ethernet-equipped introductory boards](https://digilent.com/shop/fpga-boards/development-boards/introductory-boards/?StandardExpansionConnectors=Ethernet)? 

The second question is on hold for now, as I'm lacking the desire to get too into part selection yet—e.g. none of the diligent dev gear has _two_ ethernet ports, the minimum required for an interesting router (though a single 10/100MB port would work for a ping test).

# Pinging the simulated SoC

In order to ping the SoC from my Linux host, we'd have to connect the I/O somehow from the outside world. In other contexts (e.g. containers) I've done this with a veth interface, but there's a whole [bunch of options](https://developers.redhat.com/blog/2018/10/22/introduction-to-linux-interfaces-for-virtual-networking) for "thing that can be passed to the `-I` flag of ping."

The harder challenge is probably going to be figuring out how to get the `verilator` simulation host to set itself up to translate the SoC's read attempts to interact with the `veth` (or whichever). That's going to reward understanding at least a little bit about what the IceNet "thing" is, and what I/O "looks like".

Some possible resources:

* https://chipyard.readthedocs.io/en/stable/Advanced-Concepts/Chip-Communication.html
* https://github.com/search?q=repo%3Aucb-bar%2Fchipyard%20nic&type=code
* https://groups.google.com/g/chipyard/search?q=nic
* https://docs.fires.im/en/stable/Advanced-Usage/Miscellaneous-Tips.html#experimental-support-for-sshing-into-simulated-nodes-and-accessing-the-internet-from-within-simulations

**Comment by sethp** [comment]
## Can we nic-loopback it?

Looking through the git history for `pingd.c` to see if it came with instructions, I discovered [`nic-loopback.c` test](https://github.com/ucb-bar/chipyard/blob/main/tests/nic-loopback.c) , seemingly a simpler test of the NIC simulated hardware itself.

So, after following the [Chipyard Repo Setup](https://chipyard.readthedocs.io/en/stable/Chipyard-Basics/Initial-Repo-Setup.html) docs, then:

```sh
MAKEFLAGS=-j`nproc` make -C sims/verilator VERILATOR_THREADS=1
make -C tests nic-loopback.riscv
```

followed by:

```sh
time ./sims/verilator/simulator-chipyard.harness-RocketConfig +permissive +max-cycles=10000000 +permissive-off tests/nic-loopback.riscv
```

produces this output:

```
[UART] UART0 is here (stdin/stdout).
Trial 0
*** FAILED *** (tohost = 7)
[2183535000] %Error: TestHarness.sv:166: Assertion failed in TOP.TestDriver.testHarness: Assertion failed: *** FAILED *** (exit code =          7)

    at SimTSI.scala:21 assert(!error, "*** FAILED *** (exit code = %%d)\n", exit >> 1.U)

%Error: /home/seth/Code/src/github.com/ucb-bar/chipyard/sims/verilator/generated-src/chipyard.harness.TestHarness.RocketConfig/gen-collateral/TestHarness.sv:166: Verilog $stop
Aborting...
 +permissive +max-cycles=10000000 +permissive-off tests/nic-loopback.riscv  198.70s user 0.46s system 99% cpu 3:19.57 total
```

Next ideas:

* Do some more `printf` debugging in nic-loopback.c to see if we can identify which operation is going wrong, and trace it from the source level
* Get the [JTAG debugging flow](https://chipyard.readthedocs.io/en/stable/Advanced-Concepts/Chip-Communication.html#debugging-with-jtag) up and running, so we can set a breakpoint inside `tohost` (where it's writing the 7 exit code) and work backwards from there
* Check whatever config is being generated by default to make sure it includes a NIC ; maybe "exit 7" (or 5 for `pingd.riscv`) means "no NIC found?"

**Reply by sethp** [reply]
re: 

> Check whatever config is being generated by default to make sure it includes a NIC ; maybe "exit 7" (or 5 for pingd.riscv) means "no NIC found?"

It does not: 

```scala
class RocketConfig extends Config(
  new freechips.rocketchip.subsystem.WithNBigCores(1) ++         // single rocket-core
  new chipyard.config.AbstractConfig)
```
—https://github.com/ucb-bar/chipyard/blob/3a6677bc3012f47e94b8914ebab6e789c9e949e3/generators/chipyard/src/main/scala/config/RocketConfigs.scala#L11-L13

But there is a config that seems purpose-built:
```scala
class LoopbackNICRocketConfig extends Config(
  new chipyard.harness.WithLoopbackNIC ++                      // drive NIC IOs with loopback
  new icenet.WithIceNIC ++                                     // add an IceNIC
  new freechips.rocketchip.subsystem.WithNBigCores(1) ++
  new chipyard.config.AbstractConfig)
```
—https://github.com/ucb-bar/chipyard/blob/3a6677bc3012f47e94b8914ebab6e789c9e949e3/generators/chipyard/src/main/scala/config/PeripheralDeviceConfigs.scala#L42-L46

And, luckily, it even produces a simulator with `MAKEFLAGS=-j$(nproc) make -C sims/verilator VERILATOR_THREADS=4 CONFIG=LoopbackNICRocketConfig` (many of the "stock" configs fail at the elaboration/generation steps). 

Et voila:

```
$ time /home/seth/Code/src/github.com/ucb-bar/chipyard/sims/verilator/simulator-chipyard.harness-LoopbackNICRocketConfig +permissive +max-cycles=10000000 +permissive-off tests/nic-loopback.riscv 
%Warning: System has stack size 8192 kb which may be too small; suggest 'ulimit -c 17509' or larger
[UART] UART0 is here (stdin/stdout).
Trial 0
send/recv 4396 cycles
Trial 1
send/recv 4373 cycles
Trial 2
send/recv 4382 cycles
All correct
- /home/seth/Code/src/github.com/ucb-bar/chipyard/sims/verilator/generated-src/chipyard.harness.TestHarness.LoopbackNICRocketConfig/gen-collateral/TestDriver.v:158: Verilog $finish
 +permissive +max-cycles=10000000 +permissive-off tests/nic-loopback.riscv  493.31s user 0.54s system 398% cpu 2:03.81 total
```

Figuring that out took learning to differentiate between the simulator, chipyard's different Configs, and to identify what a "Harness" and "Top" are (roughly) and quite a lot of reading various Makefiles and experimentation with `make` commands. 

It also still takes two minutes to simulate a fairly simple program, but there's plenty of avenues for exploration there too. I'm definitely attuned to the idea of diving way too deep on "make RTL fast zoom zoom," but I'm reminding myself before focusing too much on improving a single workflow, it's well worth understanding how that workflow fits into our overall goals.

**Comment by sethp** [comment]
# Can we tuntap it?

Since we're going for connecting a user-space program (the simulated risc-v chip) into the network, there's also the matter of getting the other "end" of that virtual link cable connected to something.

In Linux, "I want to connect my userspace program to a virtual network device" has a spelling of [tun/tap](https://www.kernel.org/doc/html/latest/networking/tuntap.html). They're a little awkward to use, but with help from these two resources:

* https://jvns.ca/blog/2022/09/06/send-network-packets-python-tun-tap/
* https://backreference.org/2010/03/26/tuntap-interface-tutorial/index.html

(and more than a few hours of failed experiments), I managed to `ping` a python program modified from Julia Evans' example. First:

## setup 

```bash
subnet=0
for mode in tun tap; do
sudo ip tuntap add "${mode}2" mode "${mode}" user $USER
sudo ip addr add 10.0."$(( subnet++ ))".1/24 dev "${mode}2"
sudo sysctl -w net.ipv6.conf."${mode}"2.disable_ipv6=1 
sudo ip link set "${mode}2" up
done
sudo ip neigh add 10.0.1.2 dev tap2 lladdr "$(ip --brief link show tap2 | awk '{print $3}' )" nud permanent
```

That creates a `tun2` device and `tap2` device running in tun and tap modes that are addressed by `10.0.0.1/24` and `10.0.1.1/24` respectively.[^1]

[^1]: Note: the device names are chosen arbitrarily, it's only convenient that match the pattern `{mode}{number}`, and the number 2 was chosen because you probably don't already have a `tun2` device. The subnets are a _little_ less arbitrary, IPv4 space is a whole mess, but you should probably be fine playing with the the `10.0.0.0/8` private range as long as you're running this on a local machine on a home internet connection. If your internet breaks, see `# cleanup`. If you're running on a remote machine and SSH stops working, uh, try turning it off and on again?

A `tap` device is a virtual ethernet tunnel: `write`s to the userspace end of the file descriptor cause ethernet frames to be received (RX) by the kernel for processing, as if sent by a connected peer. `read`s block until the kernel wants to transmit (TX), at which point the program wakes up with the contents of the transmitted ethernet frame in hand. 

A `tun` device is similar, but it's a pure IP tunnel; i.e. no ethernet headers are added/expected by the kernel, and it has no link-level (MAC) address.

## run it

Save this somewhere, maybe as `pingd.py`:

```python
import struct
from fcntl import ioctl

def open_dev(mode, name=b''):
    fd = open("/dev/net/tun", "r+b", buffering=0)
    LINUX_IFF_TUN = 0x0001
    LINUX_IFF_TAP = 0x0002
    LINUX_IFF_NO_PI = 0x1000
    LINUX_TUNSETIFF = 0x400454CA
    flags = LINUX_IFF_NO_PI
    if mode == 'tun':
        flags |= LINUX_IFF_TUN
    else:
        flags |= LINUX_IFF_TAP
    ifs = struct.pack("16sH22s", name, flags, b"")
    ioctl(fd, LINUX_TUNSETIFF, ifs)
    return fd

import sys
if sys.argv[1] not in ['tap', 'tun']:
    print(f'invalid mode: {sys.argv[1]}', file=sys.stderr)
    print(f'usage: {sys.argv[0]} tap|tun [DEV_NAME]', file=sys.stderr)
    sys.exit(2)

mode = sys.argv[1]
dev = open_dev(mode, (sys.argv[2:] or [''])[0].encode('utf=8'))

offset = 0 if mode == 'tun' else 14

while True:
    data = dev.read(1500)
    if len(data) == 0:
        break
    print(data.hex())
    reply = bytearray(data)

    ip_pkt = reply[offset:]
    # swap source & dest IP (doesn't change checksum)
    ip_pkt[12:16], ip_pkt[16:20] = ip_pkt[16:20], ip_pkt[12:16]
    # change from ICMP echo request (8) to ip_pkt (0)
    ip_pkt[20] = 0
    ip_pkt[22] = (ip_pkt[22] + 8) % 256 # fixup checksum
    reply[offset:] = ip_pkt
    # also swap src & dst mac addrs when we're in 'tap' mode
    if mode == 'tap':
        reply[0:6], reply[6:12] = reply[6:12], reply[0:6]
    dev.write(reply)
    print("-> ", reply.hex())
```

Then, run:

```shell
$ python pingd.py tun tun2
# or, python pingd.py tap tap2

# in another terminal
$ ping -nc2 10.0.0.2
# or, 10.0.1.2 if you ran in `tap` mode
```

And you should see familiar ping output:

```
$ ping -nc2 10.0.1.2 
PING 10.0.0.2 (10.0.0.2) 56(84) bytes of data.
64 bytes from 10.0.0.2: icmp_seq=1 ttl=64 time=0.285 ms
64 bytes from 10.0.0.2: icmp_seq=2 ttl=64 time=0.119 ms

--- 10.0.0.2 ping statistics ---
2 packets transmitted, 2 received, 0% packet loss, time 1003ms
rtt min/avg/max/mdev = 0.119/0.202/0.285/0.083 ms
```

with output similar to the following from the python side:

```
4500005424d24000400101d50a0000010a0000020800af310040000143503966000000000704060000000000101112131415161718191a1b1c1d1e1f202122232425262728292a2b2c2d2e2f3031323334353637
->  4500005424d24000400101d50a0000020a0000010000b7310040000143503966000000000704060000000000101112131415161718191a1b1c1d1e1f202122232425262728292a2b2c2d2e2f3031323334353637
45000054252f4000400101780a0000010a00000208002d25004000024450396600000000880f060000000000101112131415161718191a1b1c1d1e1f202122232425262728292a2b2c2d2e2f3031323334353637
->  45000054252f4000400101780a0000020a00000100003525004000024450396600000000880f060000000000101112131415161718191a1b1c1d1e1f202122232425262728292a2b2c2d2e2f3031323334353637
```

Which we can plug into the handy https://hpd.gasmi.net :

* https://hpd.gasmi.net/?data=4500005424D24000400101D50A0000010A0000020800AF310040000143503966000000000704060000000000101112131415161718191A1B1C1D1E1F202122232425262728292A2B2C2D2E2F3031323334353637&force=ipv4
* https://hpd.gasmi.net/?data=4500005424D24000400101D50A0000020A0000010000B7310040000143503966000000000704060000000000101112131415161718191A1B1C1D1E1F202122232425262728292A2B2C2D2E2F3031323334353637&force=ipv4

To see that we've got an ICMP echo ping request/reply.

### notes

* Don't forget to start the ping daemon before running the ping command: unless you're living in 2006 or so, your kernel will mark any `tun/tap` device with no listening userspace program as NO-CARRIER (similar to what you'd see if you yanked out an ethernet cable from your NIC right now) 
* Originally, I was trying to get `ping -I tun2 ...` to work as my user-space program without having to write a daemon _or_ mess around with the routing rules/network address space. This is ultimately doomed, I think, since ping is using the socket API: so it's not going to handle the parts of the conversation that `pingd.py` handles, and vice versa.
* It still bemuses me that it's necessary to give at least two IP addresses to the device; I guess this is because of how differently Linux thinks about a "local" address vs. "remote". That's why we have to ping `.2`, so it's routed "somewhere else" (our userspace program). It might be fun to try assigning `10.0.0.0/31` and testing with `ping 10.0.0.1`, but I'm told There Be Dragons[^dragons]
* I found it very helpful to have the python program printing out the packets it was seeing, so I could throw them into a decoder to see more exactly what was happening "on the wire". That helped a lot given that the failure mode for most of this stuff is "silently ignored"—I could at least see that _something_ was happening!
* Traffic analyzers were also useful, though they gave a more "post-processed" view of what was going on. Two similar-but-different views:
    * `sudo tcpdump -ni any icmp` (or  `sudo tcpdump -ni tun2 icmp`)
    * `sudo tshark -ni any icmp` (why does this show something different? idk!) 
* Shout out to https://baturin.org/docs/iproute2/ too: that was a helpful reference for some of the iproute2 stuff

[^dragons]: `.0` historically often had special meaning, and so not always does it work as a "plain" address (even where it ought).  

## cleanup

Stop any running python programs (lest you see `ioctl(TUNSETIFF): Device or resource busy`). Delete both devices:

```bash
for mode in tun tap; do sudo ip tuntap del "${mode}2" mode "${mode}"; done
```

## next steps ?

1. Extend the example to IPv6 so we can get out of the "oh my goodness all the subnets are colliding all the time" business; doing so may also require responding to the NDP packets that come through, and/or setting that up (statically as for `tap2` above).
2. Try to ping _out_ from the program and get a response back; this will probably require one of:
     a. Setting up NAT (ugh, nat)
     b. Messing with the kernel's routing configuration; either getting the knobs twiddled just right to respond to local traffic (which is surprisingly difficult), or doing some amount of "strategic guesswork"/DHCP to wire up an address that ~accidentally gets routed back to where it came from
     c. "Just" doing ipv6

## wrapping up / what was it all for

So, we pinged it! Now, "all" we have to do is "simply" replace our `pingd.py` with some verilog RTL simulator process!

More concretely (and the reason I've taken at least 2x as long to write this up as it did to get a little janky prototype working), this can act as a point of reference when something goes wrong in hooking up the verilog RTL simulation. If it works with `pingd.py`, but not with the simulation, then we've ruled out the whole expanse of Linux networking configuration. And, it acts as a test bed; if trying something with the simulator doesn't work, then it's possible to try the with the much-simplified `pingd.py` to see if it's possible to isolate to failure.

**Comment by sethp** [comment]
# Yes, we can! (Simulate it)

Adding this config _somewhere_ in the Chipyard Cinematic Universe (I put it in PeripheralDeviceConfigs.scala):

```scala
class TapNICRocketConfig extends Config(
  new chipyard.harness.WithSimNetwork ++
  new icenet.WithIceNIC ++
  new freechips.rocketchip.subsystem.WithNBigCores(1) ++
  new chipyard.config.AbstractConfig)
```

Reading from the bottom up (because that's how scala evaluates it), we've got:

* an AbstractConfig—that's a _lot_, but roughly it covers the "edges" of the SoC: everything from boot ROM to JTAG to simulating a UART that connects to the stdio streams of the simulator process.
* One (1) Big RocketChip Core  (i.e. a single RISC-V hart)
* An [`IceNIC`](https://chipyard.readthedocs.io/en/stable/Generators/IceNet.html), which parameterizes a memory-mapped ethernet IO peripheral on chip (very similar to the familiar esp32c3 SPI)
* A `SimNetwork` harness for the NIC (config?) that uses a linux `tap` device to send & receive packets (i.e. with read and write, where another impl might drive an external PHY)

With that in hand, we can run:

```
MAKEFLAGS=-j`nproc` make -C sims/verilator VERILATOR_THREADS=1 CONFIG=TapNICRocketConfig
```

(why `MAKEFLAGS=...` instead of just, y'know, passing the flags? because some other chipyard stuff like `./build-setup` wraps Make in a shell script that _doesn't_ forward arguments, and the env var works for). The `VERILATOR_THREADS=1` is the default and sort of unnecessary, but it's there for _foreshadowing_.

With everything [set up](https://chipyard.readthedocs.io/en/stable/Chipyard-Basics/Initial-Repo-Setup.html), that produces  a `sims/verilator/simulator-chipyard.harness-TapNICRocketConfig` binary. A little bit of setup:

```
sudo ip tuntap add mode tap dev tap0 user $USER
sudo ip link set tap0 up
sudo ip addr add 172.16.0.1/16 dev tap0
sudo sysctl -w net.ipv6.conf.tap0.disable_ipv6=1

make -C tests pingd.riscv
```

These are very similar to "can we tuntap it" above; the last bit compiles the [`pingd.c`](https://github.com/ucb-bar/chipyard/blob/main/tests/pingd.c) bare-metal RISC-V program that substitutes for `pingd.py`. The main difference between the two is that instead of using `read` & `write` directly, it uses [`nic_recv` and `nic_send`](https://github.com/ucb-bar/chipyard/blob/3a6677bc3012f47e94b8914ebab6e789c9e949e3/tests/nic.h#L29-L55) functions that write to the magic MMIO addresses the IceNIC presents as the on-device interface. Interactions with those addresses trigger the SimNetwork to invoke the underlying `read` or `write` operations.

Now we can run:

```
sims/verilator/simulator-chipyard.harness-TapNICRocketConfig +permissive +netdev=tap0 +permissive-off tests/pingd.riscv
```

Which reads: "run the simulator, letting me pass arbitrary plusargs, throw "netdev=tap0" into the void and hope SimNetwork reads it, turn strict argument parsing back on, and simulate running the bare-metal binary `tests/pingd.riscv`". NB: SimNetwork _won't_ read it without changing `SimNetwork.v` because verilator optimizes away the side-effecty read of the argument (see: https://github.com/verilator/verilator/issues/5127).

After quite a while (~80s on my machine), that'll print out something like `macaddr - 00:16:3e:7c:df:c6`, which means it's ready for `ping -i 5 172.16.0.2` from another terminal:

```
$ ping -c 10 -i 5 172.16.0.2
PING 172.16.0.2 (172.16.0.2) 56(84) bytes of data.
64 bytes from 172.16.0.2: icmp_seq=1 ttl=64 time=5318 ms
64 bytes from 172.16.0.2: icmp_seq=2 ttl=64 time=2117 ms
64 bytes from 172.16.0.2: icmp_seq=3 ttl=64 time=1873 ms
64 bytes from 172.16.0.2: icmp_seq=4 ttl=64 time=1914 ms
64 bytes from 172.16.0.2: icmp_seq=5 ttl=64 time=1934 ms
64 bytes from 172.16.0.2: icmp_seq=6 ttl=64 time=2184 ms
64 bytes from 172.16.0.2: icmp_seq=7 ttl=64 time=1899 ms
64 bytes from 172.16.0.2: icmp_seq=8 ttl=64 time=1936 ms
64 bytes from 172.16.0.2: icmp_seq=9 ttl=64 time=1854 ms
64 bytes from 172.16.0.2: icmp_seq=10 ttl=64 time=1863 ms

--- 172.16.0.2 ping statistics ---
10 packets transmitted, 10 received, 0% packet loss, time 45141ms
rtt min/avg/max/mdev = 1854.311/2289.157/5317.726/1014.939 ms, pipe 2
```

(Note the `-i` interval flag; since the default for ping is 1s, which is faster than our little simulator can process, omitting that will overwhelm the poor thing and cause it to drop so many packets the host-side Linux kernel will mark it as "unreachable")

Hooray, a 2 second ping to localhost! The first one takes a bit longer because, unlike the simpler `pingd.py`, `pingd.c` handles ARP so we didn't have to set a link-layer mapping for 172.16.0.2 with `ip neigh`, but as a consequence the simulation has to _process_ ARP, which takes a bit of additional time.

Finally, if you wanted to see the `perf -ddd` stats that got generated from that set of 10 pings there, and I know you do, here it is:

```
 Performance counter stats for '/home/seth/Code/src/github.com/ucb-bar/chipyard/sims/verilator/simulator-chipyard.harness-TapNICRocketConfig +permissive +netdev=tap0 +permissive-off tests/pingd.riscv':

        126,608.70 msec task-clock:u                     #    0.999 CPUs utilized             
                 0      context-switches:u               #    0.000 /sec                      
                 0      cpu-migrations:u                 #    0.000 /sec                      
             1,087      page-faults:u                    #    8.586 /sec                      
   529,651,994,457      cycles:u                         #    4.183 GHz                         (42.86%)
   216,906,580,340      stalled-cycles-frontend:u        #   40.95% frontend cycles idle        (42.85%)
   727,320,861,191      instructions:u                   #    1.37  insn per cycle            
                                                  #    0.30  stalled cycles per insn     (42.85%)
    88,567,784,828      branches:u                       #  699.540 M/sec                       (42.85%)
     4,490,582,640      branch-misses:u                  #    5.07% of all branches             (42.85%)
   282,884,352,466      L1-dcache-loads:u                #    2.234 G/sec                       (42.85%)
    17,706,947,853      L1-dcache-load-misses:u          #    6.26% of all L1-dcache accesses   (42.85%)
   <not supported>      LLC-loads:u                                                           
   <not supported>      LLC-load-misses:u                                                     
   163,240,550,726      L1-icache-loads:u                #    1.289 G/sec                       (42.86%)
     2,332,744,695      L1-icache-load-misses:u          #    1.43% of all L1-icache accesses   (42.85%)
       331,010,525      dTLB-loads:u                     #    2.614 M/sec                       (42.86%)
        62,437,250      dTLB-load-misses:u               #   18.86% of all dTLB cache accesses  (42.87%)
        20,475,366      iTLB-loads:u                     #  161.722 K/sec                       (42.87%)
       684,888,871      iTLB-load-misses:u               # 3344.94% of all iTLB cache accesses  (42.87%)
     8,932,705,203      L1-dcache-prefetches:u           #   70.554 M/sec                       (42.86%)
   <not supported>      L1-dcache-prefetch-misses:u                                           

     126.790467668 seconds time elapsed

     123.930254000 seconds user
       1.756772000 seconds sys
```

The real standout, of course, is the iTLB misses: `3344.94% of all iTLB cache accesses`. That's something that _would_ be helped by throwing more cores at it (since they'd all have independent iTLB caches, I believe, bringing "more cache" to bear on the problem), but there's a bug in how `SimNetwork.cc` interacts with verilator's threading model that causes a segfault in the user-land coroutine implementation when the simulator is built with any `VERILATOR_THREADS` > 1.

Potential next avenues (in increasing order of difficulty):

* Improving the ping time by fixing the bug, unlocking the power of multi-core computing (which in my testing might help with as much as a ~4x speedup, getting us to _500ms_ ping!)
* Improving the ping time by getting an FPGA and accelerating the simulator with it (this seems to be the "path of least resistance," since firesim is intended to be run in this mixed-mode, and that's probably why they haven't found the verilator bugs in IceNet)
* Getting an FPGA and figuring out what it means to build a chipyard config that "fits" it, so we can run `pingd.riscv` on it and ping _that_ target, across a physical _wire_; obviously this one needs a bit more digging, but roughly I *think* this is where we'd fit in one of these:
    * figure out how to bit-bang *MII in software, and only use the HDL for wiring up the right pins to do so
    * write [*G]MII in a HDL (possibly just MII, since right now 10 Megabit would probably feel pretty heckin' fast)

**Reply by sethp** [reply]
lol yeah, as a _very scientific_ test I tried with `ping -s 1400` to send larger packets (the default size is 56) and the ping times went up to about 24s _per packet_. Being very generous and saying the routing overhead latency doesn't matter (it absolutely does), that gives us a rough link speed for our simulator of about 61 bytes per second [(1400 - 56) / (24 -2)]; that's about 488 baud, or about 1/100th the speed of a dial-up modem. Screamin'!

**Reply by dougli1sqrd** [reply]
What is a network tap? From https://chipyard.readthedocs.io/en/stable/Generators/IceNet.html

**Reply by dougli1sqrd** [reply]
I have answered myself sort of https://en.wikipedia.org/wiki/Network_tap
Ha this sort of just introduces the question: What the heck is a tap? 

Seems like a network tap is a wire that duplicates a signal that you can read from? Why do we say "network tap" instead of "test point" or anything else?

**Reply by dougli1sqrd** [reply]
what's the diff between getting an FPGA to "accelerate" the simulator and just running the thing on the FPGA? Like if we have the FGPA  why not just put the ping/RISC-V example on there directly?

**Reply by sethp** [reply]
re: network taps; that's neat: I only knew the compound term "tap device", which means something more like "a part of the Linux kernel that emulates a physical wire (`write`s to a tap device simulate exciting the wire, and `read`s simulate sampling from it). I would guess you've found the reason it's so named, for our purposes here I think that's what IceNet is talking about. There's a bunch more about that in the "Can we tuntap it?" post if you're curious, but the first-order approximation here is that "tap in" and "tap out" in that diagram are both connected to the ping client—albeit through the socket abstraction, but now we kind of see what exactly a socket is abstracting: Ethernet frame decoding, ARP/neighbor discovery, routing, and whatever other fine details sit "above" `read`/`write` on the (pseudo-)wire.

> what's the diff between getting an FPGA to "accelerate" the simulator and just running the thing on the FPGA? Like if we have the FGPA why not just put the ping/RISC-V example on there directly?

It's a good question: one that seems like it'd let us get at some of the interesting tradeoffs to talk about. I don't really have the expertise to break it down yet, but some things I've noticed are:

* Lots of FPGAs ship so-called "hard" CPUs alongside their programmable gate array business
* Some FPGAs are way more "extra" than others, like that $10,000 datacenter-class FPGA card from #10 vs. the ~$200 RHS Research boards vs. the $14 Igloo2 (my favorite boo)
* There was a paper I saw (I can dig up the citation if you'd like) that compared various RTL simulation acceleration approaches, including their own purpose-built ASIC; IIRC the conclusion was that their ASIC outperformed FPGAs which outperformed CPUs. 

Together, those suggest to me that there's a development/production "seam" here: you build an RTL model that is intended to be turned into a bitstream for a particular shipping FPGA, but getting feedback about what's going wrong is slow and sucks, so you find a way to speed it up... and hey, you've got all this experience with FPGAs, and probably some pretty powerful ones sitting _right there_!

In other words, I think the FPGA-accelerated version is intended to be identical to the software RTL simulation (modulo the time-domain effects): it's purely deterministic, you can sample/trace any waveform without disturbing the system in any way, etc. Putting the model onto the FPGA doesn't express those same properties, but it's also probably faster/cheaper/lower power (in some sense).
